### PR TITLE
feat(ui): Refactor Project Settings and EditablePanel. Fix #4015

### LINKED ISSUE #4015: feat(ui): Improved Project summary settings
# Summary

A redesign of the Project summary settings page. 

# Motivation

In addition to improving the user experience and making it easier to update project settings, as more options are added it becomes unwieldy to merely tack a new setting to the end of a long list; some organization is necessary. By delimiting settings into sections it becomes easier to manage from both a developer and a user perspective. 

More organization than the below may be useful; for example adding shortcuts to each section at the top of the page or the ability to hide/show sections. 

# Proposal

Attached is a mockup of what I believe the updated page should look like. Additionally: 

- Fields are saved by clicking the blue button and fields that have already been saved display a green checkmark
- Settings that aren't being used are hidden or replaced with a small activation button
- To save time I did not mock up each section. The sections I've ideated are: Deployment (source repos and destinations), Allow/White List (whitelisted cluster resources and whitelisted namespace resources), Deny/Black List (blacklisted namespace resources), Signature Keys (required signature keys), and Orphaned Resources (orphaned resource monitoring and orphaned resources ignore list)

Image 7-28-20 at 5 03 PM

### PR DESCRIPTION
**Pull request by rbreeze**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Optional. My organization is added to USERS.md.
* [x] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)).

### CODE REVIEW

**Comment by rbreeze** [comment]
Features / Improvements:
- Edit settings in the same place you view them
- Access to all k8s ResourceTypes in a dropdown where appropriate, to prevent typos and ease entry
- Indication when fields are set to all (*)
- Quickly open source repo links with a convenient button
- Batch select rows in any section to delete them
- Links to documentation where relevant

**Comment by rbreeze** [comment]
Demo video: 

https://youtu.be/Sr2EeZ-ZrA4

**Review by alexmt — CHANGES_REQUESTED** [review]
Great start. I really like the idea to visually group settings and inline editing. The help text and links to the documentation is very useful too. 

The only concern is the consistency with the existing user interface:

- let's not use a separate set of colors and input field styles for the project details page. Instead, re-use the existing palette and if required update existing common colors/styles in a separate PR. 

- Argo CD already has a way to visually group collection of controls using `white-box` and `white-box_details` classes. I suggest to use it and improve if necessary, instead of inventing the new pattern:

image

Probably the best way is to try to use `EditablePanel` component since it already encapsulates switch between view and edit modes.

**Comment by sbose78** [comment]
( fixes #4015 )

**Comment by RichiCoder1** [comment]
Pardon if this is WP, and hope feedback is welcome! Had a few things I noticed:

### Header
* It looks like the revert operation doesn't work?
description bug
* It'd be nice if that field turned into either an underlined or outlined form field so it more clear it's editable.
* Does the act of editing the field set the description? If not, maybe "cancel" would be more clear than undo?
* Designer nit pick: The buttons don't seem to have hover states.

### All Groups
* I like the notes when the fields have wildcards, but they don't say anything about why I should care? Why `*` is worth calling out?
* The position and state of the checkmark doesn't say "confirm" to me, it looks like a validation artifact saying the input of that field is valid? Somewhat ok since there's a `SAVE` button at the end, but it's still confusing.

### Destinations
* This is called Cluster everywhere else, no?
image

### [Allow|Deny] List
* Ditto on feedback about allowing CRs.

### Orphaned Resources
* This is a personal bias, but magically appearing and disappearing UI can be confusing. It's especially a bit confusing in this case as it results in the contents appearing off screen. I'd say either show this UI and then gray it out, or scroll down when this UI appears so it stays in view.

off_screen_orphaned
* Other nit is that the buttons aren't in a group'd box.
* For the switches, it'd be nice if the `off` state had the `ON` text a more neutral color as otherwise it looks like these switches are actually toggled on.


That mountain of feedback out of the way, this looks like a massive step forward for this screen!

**Comment by rbreeze** [comment]
@RichiCoder1 Thank you for the feedback! This is still WIP, but your feedback is still helpful. 

With regards to the description field, editing doesn't persist the changes until `SAVE` is clicked. The intended behavior of `REVERT` is to revert the field back to what it was before the user started editing. I believe I pushed a fix for the buggy revert in the latest commit, and maybe when it works as intended it'll be more clear. 

I hadn't thought of the potential confusion with the checkmarks. Maybe a checkbox instead of 'check circles' would make that clearer? 

I think you're right that `SERVER` is called `CLUSTER` everywhere else. 

I also agree with your thoughts about the toggles. I'll explore some different options.

Thanks!

**Comment by RichiCoder1** [comment]
> I hadn't thought of the potential confusion with the checkmarks. Maybe a checkbox instead of 'check circles' would make that clearer?

Hmmm. I'm concerned that a checkbox or circle isn't the best way to communicate committed vs uncommitted inputs. 
Maybe something like this instead? 
image

**Comment by rbreeze** [comment]
the problem with this is it doesn't allow for selecting multiple line items.

**Comment by RichiCoder1** [comment]
> the problem with this is it doesn't allow for selecting multiple line items.

Ahh, didn't realize that was a goal. Is the intent for this to be table-like without looking like a table? It's a little confusing in that regard

File: ui/src/app/settings/components/project/card/card.tsx [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                            <div className='card__col-button card__col' />
+                        </div>
+                        {this.props.data.map((row, i) => {
+                            return (
+                                <div key={row.toString() + '.' + i}>
+                                    <CardRow<T>
+                                        fields={this.props.fields}
+                                        data={row}
+                                        remove={() => this.remove([i])}
+                                        save={value => this.props.save(i, value)}
```
[the remark this answers is not in the document]
Comment (rbreeze) [reply ui/src/app/settings/components/project/card/card.tsx]: Addressed by adding a new 'save all' button. Users now have the option to do either a batch save or save one line item at a time:

save-all-demo

**Comment by rbreeze** [comment]
In progress: removing Project Edit Panel altogether by streamlining project creation

**Review by alexmt — CHANGES_REQUESTED** [review]
Looks like card.scss, card.tsx almost re-implement form input [styles](https://github.com/argoproj/argo-ui/blob/master/src/styles/elements/form-controls.scss) and partially form state management. 

I've tried to implement inline editing using what we have. Converted only general settings, sources and destinations. I think result looks ok and requires significantly fewer changes: https://github.com/alexmt/argo-cd/commit/eb7e8a42db06b05dc5dcba5b8949fcb8d54fdedc

image

I agree what our UI "framework" can be improved. CSS is not pixel perfect, react-form library is abandoned etc. But I think we should work on each improvement separately. E.g. replace react-form with the successor, improve input fields colors etc.

What do you think?

**Comment by rbreeze** [comment]
@alexmt Thanks for looking at that and making those improvements - I completely agree that the improvements should be done one at a time, and that minimizing style changes in this PR is ideal. You're also right that I pretty much re-implemented the form styles, so your way is more efficient for sure.

Do you think inline editing (without having to click an "edit" button) is an improvement? I'm leaning toward yes. If not though, the last commit to this branch includes a newer version of "card" that has identical behaviour to the existing EditablePanel.

**Comment by RichiCoder1** [comment]
> Do you think inline editing (without having to click an "edit" button) is an improvement?

To somewhat butt in again, I'm camp both. Nice to have a quick and explicit way for discovery.

> But I think we should work on each improvement separately. E.g. replace react-form with the successor, improve input fields colors etc.

I do it less these days, but frontend UX is one major area I work in if there's any specific improvements or development I can assist with!

**Comment by alexmt** [comment]
The "edit" button helps up to load page quicker. Almost every edit page requires a loading list of some additional data to provide autocompletion.  So we can load view mode quicker and load autocompletion data only when the user click "Edit". 

Can you move the project details page to EditablePanel first? 

As a next step, we can work on two separate PRs for styles and for EditablePanel refactoring. This way we can ensure all edge cases are covered and new EditablePanel/Card can fully replace the existing one.

**Comment by rbreeze** [comment]
proj-refactor

**Pull request closed without merging**